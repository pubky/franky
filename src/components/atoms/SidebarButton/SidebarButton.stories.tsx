import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { fn } from 'storybook/test';
import { FileText, LockKeyhole, Users, Settings, HelpCircle, Bookmark } from 'lucide-react';

import { SidebarButton } from '@/components/atoms/SidebarButton';

const meta = {
  title: 'Atoms/SidebarButton',
  component: SidebarButton,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    icon: {
      control: false,
      description: 'Lucide icon component to display',
    },
    children: {
      control: { type: 'text' },
      description: 'Button text content',
    },
    onClick: {
      action: 'clicked',
      description: 'Click handler function',
    },
    className: {
      control: { type: 'text' },
      description: 'Additional CSS classes',
    },
  },
  args: { onClick: fn() },
} satisfies Meta<typeof SidebarButton>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    icon: FileText,
    children: 'Default Button',
  },
};

export const TermsOfService: Story = {
  args: {
    icon: FileText,
    children: 'Terms of service',
  },
};

export const PrivacyPolicy: Story = {
  args: {
    icon: LockKeyhole,
    children: 'Privacy policy',
  },
};

export const SeeAll: Story = {
  args: {
    icon: Users,
    children: 'See all',
  },
};

export const SettingsButton: Story = {
  args: {
    icon: Settings,
    children: 'Settings',
  },
};

export const HelpButton: Story = {
  args: {
    icon: HelpCircle,
    children: 'Help & Support',
  },
};

export const BookmarkButton: Story = {
  args: {
    icon: Bookmark,
    children: 'Bookmarks',
  },
};

export const WithCustomClassName: Story = {
  args: {
    icon: FileText,
    children: 'Custom Styled',
    className: 'border-red-500 hover:border-red-300',
  },
};

export const LongText: Story = {
  args: {
    icon: HelpCircle,
    children: 'This is a very long button text that might wrap',
  },
};

export const WithClickHandler: Story = {
  args: {
    icon: Settings,
    children: 'Click me',
    onClick: fn(),
  },
};

export const Disabled: Story = {
  args: {
    icon: FileText,
    children: 'Disabled Button',
    disabled: true,
  },
};

export const WithAriaLabel: Story = {
  args: {
    icon: LockKeyhole,
    children: 'Privacy',
    'aria-label': 'Open privacy policy',
  },
};

export const AllVariants: Story = {
  args: {
    icon: FileText,
    children: 'All Variants',
  },
  render: () => (
    <div className="space-y-4 w-64">
      <SidebarButton icon={FileText} onClick={fn()}>
        Terms of service
      </SidebarButton>
      <SidebarButton icon={LockKeyhole} onClick={fn()}>
        Privacy policy
      </SidebarButton>
      <SidebarButton icon={Users} onClick={fn()}>
        See all
      </SidebarButton>
      <SidebarButton icon={Settings} onClick={fn()}>
        Settings
      </SidebarButton>
      <SidebarButton icon={HelpCircle} onClick={fn()}>
        Help & Support
      </SidebarButton>
      <SidebarButton icon={Bookmark} onClick={fn()}>
        Bookmarks
      </SidebarButton>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'A collection of all common SidebarButton variants used throughout the application.',
      },
    },
  },
};
