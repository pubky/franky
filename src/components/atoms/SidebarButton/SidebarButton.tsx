import { LucideIcon } from 'lucide-react';
import * as Libs from '@/libs';

export interface SidebarButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  icon: LucideIcon;
  children: React.ReactNode;
  onClick?: () => void;
  className?: string;
}

export function SidebarButton({ icon: Icon, children, onClick, className, ...props }: SidebarButtonProps) {
  return (
    <button
      className={Libs.cn(
        'w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors',
        className,
      )}
      onClick={onClick}
      {...props}
    >
      <Icon className="h-4 w-4" />
      <span className="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]">{children}</span>
    </button>
  );
}
