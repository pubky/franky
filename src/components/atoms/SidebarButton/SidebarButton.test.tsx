import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { SidebarButton } from './SidebarButton';
import { FileText, LockKeyhole, Users } from 'lucide-react';

describe('SidebarButton', () => {
  it('renders with default props', () => {
    render(<SidebarButton icon={FileText}>Test Button</SidebarButton>);
    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('Test Button');
    expect(button).toHaveClass(
      'w-full',
      'flex',
      'items-center',
      'justify-center',
      'gap-2',
      'py-2',
      'px-4',
      'rounded-full',
      'border',
      'border-border',
      'hover:border-opacity-30',
      'cursor-pointer',
      'transition-colors',
    );
  });

  it('renders with different icons', () => {
    const { rerender } = render(<SidebarButton icon={FileText}>File Button</SidebarButton>);
    let button = screen.getByRole('button');
    expect(button.querySelector('svg')).toBeInTheDocument();

    rerender(<SidebarButton icon={LockKeyhole}>Lock Button</SidebarButton>);
    button = screen.getByRole('button');
    expect(button.querySelector('svg')).toBeInTheDocument();
    expect(button).toHaveTextContent('Lock Button');

    rerender(<SidebarButton icon={Users}>Users Button</SidebarButton>);
    button = screen.getByRole('button');
    expect(button.querySelector('svg')).toBeInTheDocument();
    expect(button).toHaveTextContent('Users Button');
  });

  it('handles click events', () => {
    const handleClick = vi.fn();
    render(
      <SidebarButton icon={FileText} onClick={handleClick}>
        Click me
      </SidebarButton>,
    );

    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('applies custom className', () => {
    render(
      <SidebarButton icon={FileText} className="custom-class">
        Custom Button
      </SidebarButton>,
    );
    const button = screen.getByRole('button');
    expect(button).toHaveClass('custom-class');
  });

  it('renders children correctly', () => {
    render(
      <SidebarButton icon={FileText}>
        <span>Custom Content</span>
      </SidebarButton>,
    );
    const button = screen.getByRole('button');
    expect(button).toHaveTextContent('Custom Content');
  });

  it('has correct icon styling', () => {
    render(<SidebarButton icon={FileText}>Icon Test</SidebarButton>);
    const icon = screen.getByRole('button').querySelector('svg');
    expect(icon).toHaveClass('h-4', 'w-4');
  });

  it('has correct text styling', () => {
    render(<SidebarButton icon={FileText}>Text Test</SidebarButton>);
    const textSpan = screen.getByRole('button').querySelector('span');
    expect(textSpan).toHaveClass('text-sm', 'font-medium');
  });

  it('forwards additional props', () => {
    render(
      <SidebarButton icon={FileText} data-testid="sidebar-button" data-custom="test" aria-label="Test button">
        Test
      </SidebarButton>,
    );
    const button = screen.getByTestId('sidebar-button');
    expect(button).toHaveAttribute('data-custom', 'test');
    expect(button).toHaveAttribute('aria-label', 'Test button');
  });

  it('can be disabled', () => {
    render(
      <SidebarButton icon={FileText} onClick={() => {}}>
        Disabled Button
      </SidebarButton>,
    );
    const button = screen.getByRole('button');
    // Note: SidebarButton doesn't have built-in disabled state, but we can test the structure
    expect(button).toBeInTheDocument();
  });
});

describe('SidebarButton - Snapshots', () => {
  it('matches snapshot with default props', () => {
    const { container } = render(<SidebarButton icon={FileText}>Default Button</SidebarButton>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with FileText icon', () => {
    const { container } = render(<SidebarButton icon={FileText}>Terms of service</SidebarButton>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with LockKeyhole icon', () => {
    const { container } = render(<SidebarButton icon={LockKeyhole}>Privacy policy</SidebarButton>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with Users icon', () => {
    const { container } = render(<SidebarButton icon={Users}>See all</SidebarButton>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with custom className', () => {
    const { container } = render(
      <SidebarButton icon={FileText} className="custom-class">
        Custom Button
      </SidebarButton>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with click handler', () => {
    const { container } = render(
      <SidebarButton icon={FileText} onClick={() => {}}>
        Clickable Button
      </SidebarButton>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with complex children', () => {
    const { container } = render(
      <SidebarButton icon={FileText}>
        <span>Complex</span> <strong>Content</strong>
      </SidebarButton>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with additional props', () => {
    const { container } = render(
      <SidebarButton icon={FileText} data-testid="test" aria-label="Test button">
        Test Button
      </SidebarButton>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
