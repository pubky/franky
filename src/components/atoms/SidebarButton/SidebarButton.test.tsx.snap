// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarButton - Snapshots > matches snapshot with FileText icon 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-file-text h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
    />
    <path
      d="M14 2v4a2 2 0 0 0 2 2h4"
    />
    <path
      d="M10 9H8"
    />
    <path
      d="M16 13H8"
    />
    <path
      d="M16 17H8"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    Terms of service
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with LockKeyhole icon 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-lock-keyhole h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="16"
      r="1"
    />
    <rect
      height="12"
      rx="2"
      width="18"
      x="3"
      y="10"
    />
    <path
      d="M7 10V7a5 5 0 0 1 10 0v3"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    Privacy policy
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with Users icon 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-users h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
    />
    <path
      d="M16 3.128a4 4 0 0 1 0 7.744"
    />
    <path
      d="M22 21v-2a4 4 0 0 0-3-3.87"
    />
    <circle
      cx="9"
      cy="7"
      r="4"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    See all
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with additional props 1`] = `
<button
  aria-label="Test button"
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
  data-testid="test"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-file-text h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
    />
    <path
      d="M14 2v4a2 2 0 0 0 2 2h4"
    />
    <path
      d="M10 9H8"
    />
    <path
      d="M16 13H8"
    />
    <path
      d="M16 17H8"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    Test Button
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with click handler 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-file-text h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
    />
    <path
      d="M14 2v4a2 2 0 0 0 2 2h4"
    />
    <path
      d="M10 9H8"
    />
    <path
      d="M16 13H8"
    />
    <path
      d="M16 17H8"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    Clickable Button
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with complex children 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-file-text h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
    />
    <path
      d="M14 2v4a2 2 0 0 0 2 2h4"
    />
    <path
      d="M10 9H8"
    />
    <path
      d="M16 13H8"
    />
    <path
      d="M16 17H8"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    <span>
      Complex
    </span>
     
    <strong>
      Content
    </strong>
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with custom className 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors custom-class"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-file-text h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
    />
    <path
      d="M14 2v4a2 2 0 0 0 2 2h4"
    />
    <path
      d="M10 9H8"
    />
    <path
      d="M16 13H8"
    />
    <path
      d="M16 17H8"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    Custom Button
  </span>
</button>
`;

exports[`SidebarButton - Snapshots > matches snapshot with default props 1`] = `
<button
  class="w-full flex items-center justify-center gap-2 py-2 px-4 rounded-full border border-border hover:border-opacity-30 cursor-pointer transition-colors"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-file-text h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
    />
    <path
      d="M14 2v4a2 2 0 0 0 2 2h4"
    />
    <path
      d="M10 9H8"
    />
    <path
      d="M16 13H8"
    />
    <path
      d="M16 17H8"
    />
  </svg>
  <span
    class="text-xs font-bold leading-normal text-[var(--base-foreground,#FFF)]"
  >
    Default Button
  </span>
</button>
`;
